### ISSUE TITLE
Add `unknown` enum to `spicy::Protocol`

### ISSUE THREAD

**Issue by ynadji** [zeek/spicy#1121]
Zeek's ports have four possible protocols: `tcp`, `udp`, `icmp`, and a catch-all `unknown`. Spicy only has the first three (and `Undef`). To improve interop with Zeek, it would be nice to add `unknown` to `spicy::Protocol`. I tried using `Undef` (both implicitly and explicitly) and it doesn't work, as I suspected. It seems worth adding a new field to differentiate between "I don't know" and "I haven't set it yet."

**Comment by rsmmr** [comment]
Conceptually, I'd say `Undef` would be the right thing to use here; it's meant to mean "something that we don't know anything more specific about". To distinguish it from "unset", one can use `optional<Protocol>`. 

That said, I think the problem here comes from the fact that the code that's converting the enum from Spicy to Zeek doesn't know that on the Zeek-side, this enum happens to have a fitting label (`unknown`) that it should map `Undef` to; it just treats it like any other enum (which means translating it to an internal value of `MAX_INT`). Not quite sure how to best address this. The three options I can see are (1) extending it on the Spicy side with an `Unknown` label, as you suggest; (2) providing a helper function for converting port that maps `Undef` to `unknown`; or (2) hardcoding ports into the current enum conversion logic to special case `Undef`->`unknown` there.

**Comment by ynadji** [comment]
Yeah, I like `Undef` as long as it can be differentiated between "unset" so that's cool. (2) and (3) both seem fine to me and in my very inexperienced opinion, it seems like a specific enough case to do (3). If this is as straightforward as it seems to me, I'd be happy to tackle this as a first issue if you could give some pointers :).

**Comment by rsmmr** [comment]
Yeah, mulling over this a bit more, I prefer (3), too. Seems like some nice little magic to have, even if it's a bit odd to hardcode that one type. 

If you're game to give it a try, the conversion function from Spicy enum to Zeek enum is here: https://github.com/zeek/spicy-plugin/blob/13d822c3ede6bfab79a7ffe57aa74f1e3ef830f8/include/runtime-support.h#L632

So when the Spicy value is < 0, it's `Undef`; then we'd need to see if the Zeek-side type is `port` and use the corresponding value instead of int::max.
